packages/rest/src/lib/RequestManager.ts

```diff
@@ -0,0 +1,280 @@
+
+	// eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+	#token: string | null = null;
+
+	public constructor(public readonly options: RESTOptions) {
+		super();
+	}
+
+	public setToken(token: string) {
+		this.#token = token;
+		return this;
+	}
+
+	public async queueRequest(request: InternalRequest): Promise<unknown> {
+		// Generalize the endpoint to its route data
+		const routeID = RequestManager.generateRouteData(request.fullRoute, request.method);
+		// Get the bucket hash for the generic route, or point to a global route otherwise
+		const hash =
+			this.hashes.get(`${request.method}:${routeID.bucketRoute}`) ?? `Global(${request.method}:${routeID.bucketRoute})`;
+
+		// Get the request handler for the obtained hash, with its major parameter
+		const handler =
+			this.handlers.get(`${hash}:${routeID.majorParameter}`) ?? this.createHandler(hash, routeID.majorParameter);
+
+		// Resolve the request into usable fetch/node-fetch options
+		const { url, fetchOptions } = this.resolveRequest(request);
+
+		// Queue the request
+		return handler.queueRequest(routeID, url, fetchOptions);
+	}
+
+	/**
+	 * Creates a new rate limit handler from a hash, based on the hash and the major parameter
+	 * @param hash The hash for the route
+	 * @param majorParameter The major parameter for this handler
+	 * @private
+	 */
+	private createHandler(hash: string, majorParameter: string) {
+		// Create the async request queue to handle requests
+		const queue = new SequentialHandler(this, hash, majorParameter);
+		// Save the queue based on its ID
+		this.handlers.set(queue.id, queue);
+
+		return queue;
+	}
+
+	/**
+	 * Formats the request data to a usable format for fetch
+	 * @param request The request data
+	 */
+	private resolveRequest(request: InternalRequest): { url: string; fetchOptions: RequestInit } {
+		const { options } = this;
+
+		let query = '';
+
+		// If a query option is passed, use it
+		if (request.query) query = `?${request.query.toString() as string}`;
+
+		// Create the required headers
+		const headers: RequestHeaders = {
+			'User-Agent': `${DefaultUserAgent} ${options.userAgentAppendix}`.trim(),
+		};
+
+		// If this request requires authorization (allowing non-"authorized" requests for webhooks)
+		if (request.auth !== false) {
+			// If we haven't received a token, throw an error
+			if (!this.#token) throw new Error('Expected token to be set for this request, but none was present');
+
+			headers.Authorization = `${request.authPrefix ?? 'Bot'} ${this.#token}`;
+		}
+
+		// If a reason was set, set it's appropriate header
+		if (request.reason?.length) {
+			headers['X-Audit-Log-Reason'] = encodeURIComponent(request.reason);
+		}
+
+		// Format the full request URL (api base, optional version, endpoint, optional querystring)
+		const url = `${options.api}${request.versioned === false ? '' : `/v${options.version}`}${
+			request.fullRoute
+		}${query}`;
```

Comment:
```suggestion
		const url = `${options.api}${request.versioned ? `/v${options.version}` : ''}${request.fullRoute}${query}`;
```

Reply:
N/A, done by our lint rules

Reply:
The explicit `=== false` is intentional, see https://github.com/discordjs/discord.js-next/pull/34#discussion_r565669214 for reference.